src/merkle_tree/archive_backend.cpp

```diff
@@ -0,0 +1,92 @@
+#include <morum/archive_backend.hpp>
+
+#include <qtils/macro/unwrap.hpp>
+
+namespace morum {
+
+  std::expected<std::optional<TreeNode>, StorageError> ArchiveNodeLoader::load(
+      qtils::BitSpan<>, const Hash32 &hash) const {
+    Hash32 hash_copy = hash;
+    hash_copy[0] &= 0xFE;
+    qtils::ByteArr<sizeof(Leaf)>
+        node_data;  // in-memory branches contain non-serializable metadata,
+                    // so Leaf's size is used
+    QTILS_UNWRAP(auto res, node_storage->read_to(hash_copy, node_data));
+    if (!res) {
+      return std::nullopt;
+    }
+    QTILS_ASSERT_EQ(*res, sizeof(Leaf));
+    return deserialize_node(node_data);
+  }
+
+  ArchiveTrieDb::ArchiveTrieDb(
+        std::shared_ptr<ColumnFamilyStorage<ColumnFamilyId>> storage)
+        : storage_{storage},
+          node_storage_{storage_->get_column_family(ColumnFamilyId::TREE_NODE)},
+          value_storage_{storage_->get_column_family(ColumnFamilyId::TREE_VALUE)} {
+      QTILS_ASSERT(storage_ != nullptr);
+      QTILS_ASSERT(node_storage_ != nullptr);
+      QTILS_ASSERT(value_storage_ != nullptr);
+    }
+
+    std::expected<std::optional<std::unique_ptr<MerkleTree>>, StorageError>
+    ArchiveTrieDb::load_tree(const Hash32 &root_hash) const {
+      auto root_bytes_res = node_storage_->read(root_hash);
+      QTILS_ASSERT_HAS_VALUE(root_bytes_res);
+      QTILS_ASSERT(root_bytes_res->has_value());
+      auto root_bytes = **root_bytes_res;
+      auto root = morum::deserialize_node(root_bytes);
+      return std::make_unique<MerkleTree>(root,
+          std::make_unique<FlatPagedNodeStorage>(),
+          std::make_shared<ArchiveNodeLoader>(node_storage_));
+    }
+
+    std::unique_ptr<MerkleTree> ArchiveTrieDb::empty_tree() const {
+      return std::make_unique<MerkleTree>(
+          std::make_unique<FlatPagedNodeStorage>(),
+          std::make_shared<NoopNodeLoader>());
+    }
+
+    std::expected<Hash32, StorageError> ArchiveTrieDb::get_root_and_store(
+        const MerkleTree &tree) {
+      auto batch = storage_->start_batch();
+
+      auto hash = tree.calculate_hash([&](const morum::TreeNode &n,
+                                          qtils::ByteView serialized,
+                                          qtils::ByteView hash,
+                                          qtils::BitSpan<>) {
+        morum::Hash32 hash_copy;
+        std::ranges::copy(hash, hash_copy.begin());
+        hash_copy[0] &= 0xFE;
+        [[maybe_unused]] auto res =
+            batch->write(ColumnFamilyId::TREE_NODE, hash_copy, serialized);
+        QTILS_ASSERT(res);
+        if (n.is_leaf()) {
+          // cached_nodes_.emplace(hash_copy, n);
+        } else {
```

Comment:
```suggestion
        if (not n.is_leaf()) {
```
or uncomment. Or refactor.